types/types.go

```diff
@@ -209,3 +183,29 @@ type Metrics struct {
 	// Cumulative size of all elements in memory cache (in bytes)
 	SizeMemoryCache uint64
 }
+
+// Array is a wrapper around a slice that ensures that we get "[]" JSON for nil values
+type Array[C any] []C
+
+// MarshalJSON ensures that we get "[]" for nil arrays
+func (a Array[C]) MarshalJSON() ([]byte, error) {
+	if len(a) == 0 {
+		return []byte("[]"), nil
+	}
+	var raw []C = a
+	return json.Marshal(raw)
+}
+
+// UnmarshalJSON ensures that we get nil for "[]" and "null"
+func (a *Array[C]) UnmarshalJSON(data []byte) error {
+	// make sure we deserialize [] back to null
+	if string(data) == "[]" || string(data) == "null" {
```

Comment:
Not sure I get the point here, but we'd get `null` if the contract sent something like `"msg_responses": null`. We don't really need the extra check here though, since that would automatically result in a `nil` when deserializing below.